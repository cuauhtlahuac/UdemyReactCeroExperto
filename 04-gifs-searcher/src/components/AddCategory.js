import React, { useState } from 'react';
import PropTypes from 'prop-types';

function AddCategory({
	setCategories,
	setCategoryDictionary,
	setSameValue,
	categoryDictionary,
}) {
	const [ inputValue, setInputValue ] = useState('');

	const handleInputValue = e => setInputValue(e.target.value);

	const handleSubmit = e => {
		e.preventDefault();

		const trimmedInputValue = inputValue.trim();

		if (trimmedInputValue && trimmedInputValue.length > 1) {
			setCategoryDictionary(catDic => ({
				...catDic,
				[trimmedInputValue]: trimmedInputValue,
			}));

			if (categoryDictionary.hasOwnProperty(trimmedInputValue)) {
				setSameValue(true);
			} else {
				setCategories(cat => [ trimmedInputValue, ...cat ]);
			}
			setInputValue('');
		}
	};

	return (
		<form onSubmit={handleSubmit} className="form">
		<label htmlFor={"category"}>Category</label>
			<input
				type="text"
				id={"category"}
				value={inputValue}
				placeholder="write a favorite gif category"
				onChange={handleInputValue}
			/>
		</form>
	);
}

export default AddCategory;

AddCategory.propTypes = {
	setCategories: PropTypes.func.isRequired,
};
