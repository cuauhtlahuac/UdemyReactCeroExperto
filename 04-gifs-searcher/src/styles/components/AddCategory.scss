form {
	font-family: 'open sans', Arial, Helvetica, sans-serif;
	display: flex;
	flex-direction: column;
	margin-top: 1.3rem;
}

form > label {
	margin-bottom: 5px;

animation-duration: 5s;
animation-iteration-count: infinite;
animation-name: label-color;
}

@keyframes label-color {
	from {
		color: white;
	}

	to {
		color: black;
	}
}