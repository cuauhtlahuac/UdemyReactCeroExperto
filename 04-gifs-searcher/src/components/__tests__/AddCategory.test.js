import React from 'react';
import "@testing-library/jest-dom";
import { shallow } from 'enzyme';
import AddCategory from 'components/AddCategory';

describe('AddCategory tests', () => {
    const setCategories =   jest.fn();
    const setCategoryDictionary = jest.fn();
    const setSameValue = jest.fn();
    
    let component = shallow( <p /> );

    beforeEach(() => {

        component = shallow( 
        <AddCategory 
        setCategories={ setCategories } 
        setCategoryDictionary={ setCategoryDictionary }
        categoryDictionary={{}}
        setSameValue={ setSameValue }
        />
        );
        
    })
    
	test('should render correctly', () => {

		expect(component).toMatchSnapshot();
 
    });

    test('should change input value', ( ) => {
        
        const input = component.find('input');
        const value = 'Hello ðŸŒŽ';

        input.simulate('change', { target: { id: "category", value } });
        
        expect( component.find('input').prop('value') ).toBe( value );
        expect( component ).toMatchSnapshot();
    });

    test('should not call setCategories with empty value', () => {

        component.find('form').simulate("submit", {preventDefault(){}});

        expect( setCategories ).not.toHaveBeenCalled();
    });

    test('should not call setCategoryDictionary with empty value', () => {

        component.find('form').simulate("submit", {preventDefault(){}});

        expect( setCategoryDictionary ).not.toHaveBeenCalled();
    });

    test('should not call setSameValue with empty value', () => {
        
        component.find('form').simulate("submit", {preventDefault(){}});

        expect( setSameValue ).not.toHaveBeenCalled();
    });

    test('should call setCategories and clean text box', () => {
        
        component.find('input').simulate('change',{ target: { value: 'Lets play ðŸ•¹' } });
        
        expect( component ).toMatchSnapshot();
        
        component.find( 'form' ).simulate("submit", {preventDefault(){}});

        expect( component ).toMatchSnapshot();
        expect( setCategories ).toHaveBeenCalledTimes(1);
        expect( setCategories ).toHaveBeenCalledWith( expect.any( Function ) ); // Para verificar que fue llamado con una funciÃ³n
        expect( component.find( 'input' ).prop( 'value' ) ).toBe( '' );

    });
    
    test('should call setCategoryDictionary once', () => {

        expect( setCategoryDictionary ).toHaveBeenCalledTimes(1);
    });

    test('should call setSameValue when the value has repeated', () => {
        const value = 'duplicated value ðŸ‘¥'

        const component = shallow( 
            <AddCategory 
            setCategories={ setCategories } 
            setCategoryDictionary={ setCategoryDictionary }
            categoryDictionary={{ [value]: value }}
            setSameValue={ setSameValue }
            />
        );

        component.find('input').simulate('change',{ target: { value } });
        component.find( 'form' ).simulate("submit", {preventDefault(){}});

        expect( setSameValue ).toHaveBeenCalledTimes(1);
    });    
});
