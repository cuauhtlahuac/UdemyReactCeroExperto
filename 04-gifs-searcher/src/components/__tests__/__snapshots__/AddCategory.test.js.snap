// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCategory tests should call setCategories and clean text box 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <label
    htmlFor="category"
  >
    Category
  </label>
  <input
    id="category"
    onChange={[Function]}
    placeholder="write a favorite gif category"
    type="text"
    value="Lets play ðŸ•¹"
  />
</form>
`;

exports[`AddCategory tests should call setCategories and clean text box 2`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <label
    htmlFor="category"
  >
    Category
  </label>
  <input
    id="category"
    onChange={[Function]}
    placeholder="write a favorite gif category"
    type="text"
    value=""
  />
</form>
`;

exports[`AddCategory tests should change input value 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <label
    htmlFor="category"
  >
    Category
  </label>
  <input
    id="category"
    onChange={[Function]}
    placeholder="write a favorite gif category"
    type="text"
    value="Hello ðŸŒŽ"
  />
</form>
`;

exports[`AddCategory tests should render correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <label
    htmlFor="category"
  >
    Category
  </label>
  <input
    id="category"
    onChange={[Function]}
    placeholder="write a favorite gif category"
    type="text"
    value=""
  />
</form>
`;
